mir_perception/mir_empty_space_detection/ros/src/empty_space_detector.cpp

```diff
@@ -0,0 +1,268 @@
+
+    nh_.param<float>("empty_space_point_count_percentage_threshold", empty_space_pnt_cnt_perc_thresh_, 0.8);
+    nh_.param<float>("empty_space_radius", empty_space_radius_, 0.05);
+    expected_num_of_points_ = (empty_space_radius_ * empty_space_radius_ * 3.14159) / (voxel_leaf_size * voxel_leaf_size);
+}
+
+void EmptySpaceDetector::eventInCallback(const std_msgs::String::ConstPtr &msg)
+{
+    std_msgs::String event_out;
+    if (msg->data == "e_add_cloud")
+    {
+        add_to_octree_ = true;
+        return;
+    }
+    else if (msg->data == "e_add_cloud_stop")
+    {
+        add_to_octree_ = false;
+        event_out.data = "e_add_cloud_stopped";
+    }
+    else if (msg->data == "e_trigger")
+    {
+        bool success = findEmptySpaces();
+        event_out.data = (success) ? "e_success" : "e_failure";
+    }
+    else
+    {
+        return;
+    }
+    event_out_pub_.publish(event_out);
+}
+
+void EmptySpaceDetector::pcCallback(const sensor_msgs::PointCloud2::ConstPtr &msg)
+{
+    if (add_to_octree_)
+    {
+        sensor_msgs::PointCloud2 msg_transformed;   
+        if (!mpu::pointcloud::transformPointCloudMsg(tf_listener_, output_frame_, *msg, msg_transformed))
+            return;
+        
+        PointCloud::Ptr input_pc(new PointCloud);
+        pcl::fromROSMsg(msg_transformed, *input_pc);
+
+        cloud_accumulation_->addCloud(input_pc);
+        
+        add_to_octree_ = false;
+
+        std_msgs::String event_out;
+        event_out.data = "e_added_cloud";
+        event_out_pub_.publish(event_out);
+    }
+}
+
+bool EmptySpaceDetector::findEmptySpaces()
+{
+    PointCloud::Ptr plane(new PointCloud);
+    bool plane_found = this->findPlane(plane);
+    if (!plane_found)
+    {
+        return false;
+    }
+
+    geometry_msgs::PoseArray empty_space_poses;
+    this->findEmptySpacesOnPlane(plane, empty_space_poses);
+    ROS_DEBUG_STREAM(empty_space_poses);
+    if (empty_space_poses.poses.size() == 0)
+    {
+        return false;
+    }
+
+    pose_array_pub_.publish(empty_space_poses);
+
+    if (enable_debug_pc_pub_)
+    {
+        /* publish debug pointcloud */
+        sensor_msgs::PointCloud2 output;
+        pcl::toROSMsg(*plane, output);
+        output.header.frame_id = output_frame_;
+        output.header.stamp = ros::Time::now();
+        pc_pub_.publish(output);
+        ROS_INFO("Publishing debug pointcloud");
+    }
+
+    return true;
+}
+
+void EmptySpaceDetector::findEmptySpacesOnPlane(const PointCloud::Ptr &plane,
+                                                geometry_msgs::PoseArray &empty_space_poses)
+{
+    srand(time(NULL));
+    int num_of_empty_spaces_required;
+    nh_.param<int>("num_of_empty_spaces_required", num_of_empty_spaces_required, 3);
+    empty_space_poses.header.frame_id = output_frame_;
+    empty_space_poses.header.stamp = ros::Time::now();
+    int attempts = 0;
+
+    while (attempts < 100 * num_of_empty_spaces_required)
```

Comment:
make the 100 configurable.
Isint it better to do this basis on time . like search for 3 secs.?
Maybe add a ros topic in which the user can change the amount of time the user is ready to wait for an answer .

Reply:
I've made the duration for attempt (`trial_duration` in code) as a ros parameter. This felt most natural since this is a parameter for calculation but not information needed for it. One can simply  set parameter (`rospy.set_parameter` in python) and then send `e_trigger` to have the same effect as you mentioned.